src/data/finsupp/to_dfinsupp.lean

```diff
@@ -187,4 +190,70 @@ noncomputable def finsupp_lequiv_dfinsupp
   map_add' := finsupp.to_dfinsupp_add,
   .. finsupp_equiv_dfinsupp}
 
+section sigma
+/-- ### Stronger versions of `finsupp.split` -/
+
+noncomputable theory
+open_locale classical
+
+variables {η : ι → Type*} {N : Type*} [semiring R]
+
+open finsupp
+
+/-- `finsupp.split` is an equivalence between `(Σ (j : η), ιs j) →₀ α`
+and `Π₀ j, (ιs j →₀ α)`. -/
+def sigma_finsupp_equiv_dfinsupp [has_zero N] : ((Σ i, η i) →₀ N) ≃ (Π₀ i, (η i →₀ N)) :=
+{ to_fun := λ f, ⟦⟨split f, (split_support f : finset ι).val, λ i,
+    begin
+    rw [← finset.mem_def, mem_split_support_iff_nonzero],
+    exact (decidable.em _).symm
+    end⟩⟧,
+  inv_fun := λ f,
+  begin
+    refine on_finset (finset.sigma f.support (λ j, (f j).support)) (λ ji, f ji.1 ji.2)
+      (λ g hg, finset.mem_sigma.mpr ⟨_, mem_support_iff.mpr hg⟩),
+    simp only [ne.def, dfinsupp.mem_support_to_fun],
+    intro h,
+    rw h at hg,
+    simpa using hg
+  end,
+  left_inv := λ f, by { ext, simp [split] },
+  right_inv := λ f, by { ext, simp [split] } }
+
+@[simp]
+lemma sigma_finsupp_equiv_dfinsupp_apply [has_zero N] (f : Π₀ i, (η i →₀ N)) (s : Σ i, η i) :
+  (sigma_finsupp_equiv_dfinsupp.symm f : (Σ i, η i) →₀ N) s = f s.1 s.2 := rfl
+
+@[simp]
+lemma sigma_finsupp_equiv_dfinsupp_symm_apply [has_zero N] (f : (Σ i, η i) →₀ N) :
+  (sigma_finsupp_equiv_dfinsupp f : Π i, (η i →₀ N)) = finsupp.split f := rfl
+
+-- Without this Lean fails to find the `add_zero_class` instance on `Π₀ i, (η i →₀ N)`.
+local attribute [-instance] finsupp.has_zero
+
+/-- `finsupp.split` is an additive equivalence between `(Σ (j : η), ιs j) →₀ α`
+and `Π₀ j, (ιs j →₀ α)`. -/
+def sigma_finsupp_add_equiv_dfinsupp [add_zero_class N] : ((Σ i, η i) →₀ N) ≃+ (Π₀ i, (η i →₀ N)) :=
+{ map_add' := λ f g,
+  begin
+    ext,
+    simpa [finsupp.split_apply],
+  end,
+  .. sigma_finsupp_equiv_dfinsupp }
+
+local attribute [-instance] finsupp.add_zero_class finsupp.add_monoid
+
+/-- `finsupp.split` is a linear equivalence between `(Σ (j : η), ιs j) →₀ α`
+and `Π₀ j, (ιs j →₀ α)`. -/
+def sigma_finsupp_lequiv_dfinsupp [add_comm_monoid N] [module R N] :
+  ((Σ i, η i) →₀ N) ≃ₗ[R] (Π₀ i, (η i →₀ N)) :=
+{ map_smul' := λ r f,
+  begin
+    ext,
+    simpa [finsupp.split_apply],
+  end,
```

Comment:
I think you should extract both this `map_add'` and the `map_smul` belong into standalone simp lemmas, `sigma_finsupp_equiv_dfinsupp_add` and `sigma_finsupp_equiv_dfinsupp_smul`.

Reply:
I am trying to prove
```lean
lemma sigma_finsupp_equiv_dfinsupp_add [add_zero_class N] (f g : (Σ i, η i) →₀ N) :
  (sigma_finsupp_equiv_dfinsupp (f + g) : Π₀ i, (η i →₀ N)) = (sigma_finsupp_equiv_dfinsupp f) +
  (sigma_finsupp_equiv_dfinsupp g) :=
```
But I get a `failed to synthesize type class instance for` error: Lean doesn't find `has_add (Π₀ (i : ι), η i →₀ N)`. This is the same error I got in `sigma_finsupp_add_equiv_dfinsupp` and that was solved by `local attribute [-instance] finsupp.has_zero`. I have no idea what is going on.

Reply:
Note that of course this is literally `map_add'` in `sigma_finsupp_add_equiv_dfinsupp`...

Reply:
I think you should use the same trick of `local attribute [-instance]` when stating that lemma for now.

Reply:
I mean that it doesn't work even with that.
```lean
local attribute [-instance] finsupp.has_zero
--doesn't work
lemma sigma_finsupp_equiv_dfinsupp_add [add_monoid N] (f g : (Σ i, η i) →₀ N) :
  (sigma_finsupp_equiv_dfinsupp (f + g) : Π₀ i, (η i →₀ N)) = (sigma_finsupp_equiv_dfinsupp f) +
  (sigma_finsupp_equiv_dfinsupp g) :=

--works
/-- `finsupp.split` is an additive equivalence between `(Σ i, η i) →₀ N` and `Π₀ i, (η i →₀ N)`. -/
@[simps]
def sigma_finsupp_add_equiv_dfinsupp [add_zero_class N] : ((Σ i, η i) →₀ N) ≃+ (Π₀ i, (η i →₀ N)) :=
{ to_fun := sigma_finsupp_equiv_dfinsupp,
  inv_fun := sigma_finsupp_equiv_dfinsupp.symm,
  map_add' := λ f g, by {ext, refl},
  .. sigma_finsupp_equiv_dfinsupp }
```

Reply:
Why `add_monoid` and not `add_zero_class` in that first example?

Reply:
Sorry, I was experimenting. It is the same with `add_zero_class`.